ipyvizzu.py

```diff
@@ -205,10 +247,22 @@ def _validate(self, animation):
         common_keys = set(data).intersection(self)
 
         if common_keys:
+            if common_keys == {"data"}:
+                return self._validate_data(data)
```

Comment:
I'm just thinking, but wouldn't it be a cleaner solution to extend `Data` with `add_filter` method? So we could avoid the extra logic here.

```python
class Data(dict, Animation):
    def add_filter(self, expression: str):
        self["filter"] = RawJavaScript(f"record => {{ return ({expression}) }}")
```

Even the `Filter` class may could be skipped.

Reply:
I created the Filter class because the most common usecase is to set filters without any other data (data set in previous steps).

We were talking about it with @simzer and an another idea was similar to your example above,
- add add_filter to Data class, but create a Filter class and in the constructor create a Data class and call add_filter, but the anim merging in that case was still problematic, and we can add filter in different classes too.
- If only the Data class has an add_filter, I think it it less readable (from the user's point of view) in the next steps, when the user needs to create an empty Data class to add a filter.
- So I created _validate_data, and separate Data and Filter so the user can't add filter in Data, but only with Filter.
I am really not sure is it the best solution.

@nyirog Do you think the validate_data causes unacceptable overhead, and what do you think about the about the above options?

Reply:
No, it's not a big issue.

My first hunch was if the `filter` will be passed under `data` why shouldn't we handle `filter` with the `Data` python class?

On the javascript side the user writes:
```javascript
chart.animate({
    data: {
        filter: null
    }
})
```

Therefore I thought that the extra `Data()` prefix would follow the javascript API:
```python
cart.animate(Data().add_filter(None))
```
However the `add_filter()` maybe not a good method name. It could be simply `filter()`.

Or `Data` could accept `Filter` as constructor parameter. And maybe we can extend the `Data` constructor parameter options with `Record`, `Series`, `Dimension` and `Measure` later on. In this way the class names would follow the js API.

`ipyvizzu` is a javascript code generator so I think it should be as close as possible to the js API. But I'm not so familiar with `vizzu` itself. I think you have a better overview of it so do as you think it is fits better.

Reply:
I think, if it could be used this way:
```python
chart.animate(Data(Filter(None))
```
than @nyirog suggestion would be a good trade off, keeping the API closer to the original.